File: Advanced_Robotics_Course/Chapter3/Lesson1/Chapter3_Lesson1.java
import java.util.Random;

public class PCMonteCarlo {
    private static final Random rng = new Random();

    public static boolean trialSuccess(int numSamples, int dim, double radius) {
        double[] center = new double[dim];
        for (int i = 0; i < dim; ++i) {
            center[i] = 0.5;
        }
        for (int i = 0; i < numSamples; ++i) {
            double dist2 = 0.0;
            for (int d = 0; d < dim; ++d) {
                double x = rng.nextDouble();
                double diff = x - center[d];
                dist2 += diff * diff;
            }
            if (Math.sqrt(dist2) <= radius) {
                return true;
            }
        }
        return false;
    }

    public static double estimateSuccessProb(int numSamples,
                                             int dim,
                                             double radius,
                                             int trials) {
        int successes = 0;
        for (int t = 0; t < trials; ++t) {
            if (trialSuccess(numSamples, dim, radius)) {
                successes++;
            }
        }
        return ((double) successes) / trials;
    }

    public static void main(String[] args) {
        int[] ns = {10, 50, 100, 200, 500, 1000};
        for (int n : ns) {
            double p = estimateSuccessProb(n, 4, 0.1, 1000);
            System.out.println("n = " + n +
                               ", approx P(success) = " + p);
        }
    }
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter3/Lesson1/Chapter3_Lesson1.cpp
#include <iostream>
#include <random>
#include <cmath>

bool trial_success(int numSamples, int dim, double radius) {
    static thread_local std::mt19937 gen(std::random_device{}());
    std::uniform_real_distribution<double> unif(0.0, 1.0);

    // center at 0.5 in each coordinate
    std::vector<double> center(dim, 0.5);

    for (int i = 0; i < numSamples; ++i) {
        double dist2 = 0.0;
        for (int d = 0; d < dim; ++d) {
            double x = unif(gen);
            double diff = x - center[d];
            dist2 += diff * diff;
        }
        if (std::sqrt(dist2) <= radius) {
            return true; // at least one sample hits the goal tube
        }
    }
    return false;
}

double estimate_success_prob(int numSamples, int dim = 4,
                             double radius = 0.1, int trials = 1000) {
    int successCount = 0;
    for (int t = 0; t < trials; ++t) {
        if (trial_success(numSamples, dim, radius)) {
            ++successCount;
        }
    }
    return static_cast<double>(successCount) / trials;
}

int main() {
    for (int n : {10, 50, 100, 200, 500, 1000}) {
        double p = estimate_success_prob(n);
        std::cout << "n = " << n
                  << ", approx P(success) = " << p << std::endl;
    }
    return 0;
}
      
